### DISCUSSION TITLE
Add Kubernetes support

### DISCUSSION THREAD

**Discussion by Sporesirius** [Lodestone-Team/lodestone#138] [Ideas]
Kubernetes support for web, core, and game servers.
Isolation of game servers in containers.

**Comment by pinnouse** [comment]
Could you expand more on how you'd like to see this support?

**Reply by Sporesirius** [reply]
Hi @pinnouse, sorry for taking so long to reply.

The ability to host game server in Kubernetes e.g auto scaling and load balance instances of multiple game servers, but also hosting the Lodestone web interface and core daemon in Kubernetes.

There is an interesting Kubernetes project from Google that might can be the basis for this feature request:
https://github.com/googleforgames/agones

**Reply by pinnouse** [reply]
As a first party support, it may not be something we have in the pipeline just yet, but this type of functionality may be supported by upcoming macros functionality

**Comment by tropics1999** [comment]
I want to second this, I host Lodestone on Kubernetes at the moment and think there is a lot of scope for a really neat product here, with little to no competition on the "enthusiast home user" side of the spectrum. Personally I care little for the auto-scaling or high availability side of Kubernetes for this use case (I imagine the game servers would have to support this, which many would not), and this is already serviced - for Minecraft at least - with Shulker. However, the multi-node capabilities of Kubernetes are what led me to it, and I think is where integration with Lodestone would really shine. 

Effectively, I envisage this would be a similar capability to the one you are building with Docker at the moment, where Lodestone interacts with and can manage containers running on the cluster. Even if I had to add the deployments manually, and then could manage them through Lodestone I would be happy.

To draw parallels with an existing app, Coder does this in a really effective way currently, where you define a "Template" that then creates a "Workspace" on the cluster. The app Homepage also interacts with the Kubernetes API's, if in a more simplistic manner, to show if a service is running and some basic node/cluster statistics.

Furthermore, this would be a nice differentiation from other game hosting applications like Pterodactyl, where they effectively duplicate the multi-node capabilities of Kubernetes with their Wings system - which is not ideal if you want to host anything that is not a game server across those same nodes (in my opinion at least). 

TL;DR; I think Lodestone should leverage Kubernetes for managing game servers across multiple nodes, by interacting with it, the same way it does with Docker on a single node.

**Reply by pinnouse** [reply]
So if I'm getting this right, it would be nice to monitor + manage other nodes within the Kubernetes cluster that your lodestone instance is a part of, i.e. by leveraging the Kubernetes API?
